File: 이것이코딩테스트다/BFSDFS/음료수얼려먹기.py
N, M = map(int, input().split())

graph = []
for i in range(N):
  graph.append(list(map(int, input())))

visited = [[False] * M for _ in range(N)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def dfs(x, y):
  # 그래프 범위를 벗어나거나, 방문했거나, 얼음이 아닌 경우
  if x < 0 or x >= N or y < 0 or y >= M or visited[x][y] or graph[x][y] == 1:
    return

  visited[x][y] = True
  
  # 상하좌우 탐색
  dfs(x - 1, y)
  dfs(x + 1, y)
  dfs(x, y - 1)
  dfs(x, y + 1)


answer = 0
for x in range(N):
  for y in range(M):
    if graph[x][y] == 0 and not visited[x][y]:
      dfs(x, y)
      answer += 1

print(answer)

File: 이것이코딩테스트다/BFSDFS/음료수얼려먹기.js
const fs = require('fs')
const input = fs.readFileSync('input.txt').toString().trim().split('\n');

let [N, M] = input[0].split(' ').map(Number)

let graph = []
for (let i = 1; i <= N; i++) {
  graph.push(input[i].split('').map(Number))
}
let dx = [0, 0, 1, -1]
let dy = [-1, 1, 0, 0]

// 2. DFS 활용

let visited = Array.from({ length: N }, () => Array(M).fill(false));
let answer = 0

const dfs = (x, y) => {
  if (x < 0 || x >= N || y < 0 || y >= M || visited[x][y] || graph[x][y] === 1) {
    return
  }
  visited[x][y] = true
  dfs(x - 1, y)
  dfs(x + 1, y)
  dfs(x, y - 1)
  dfs(x, y + 1)
}

answer = 0
for (let x = 0; x < N; x++) {
  for (let y = 0; y < M; y++) {
    if (graph[x][y] == 0 && !visited[x][y]) {
      dfs(x, y)
      answer += 1
    }
  }
}

console.log(answer)